### ISSUE TITLE
FID evaluation reference data

### ISSUE THREAD

**Issue by MArSha1147** [LTH14/mar#8]
Thank you for your great work.

I am curious about why adm_in256_stats.npz generated by [guided-diffusion](https://github.com/openai/guided-diffusion/tree/main/evaluations), is used as the label for ImageNet 256x256 FID evaluation, rather than the [standard reference data](https://openaipublic.blob.core.windows.net/diffusion/jul-2021/ref_batches/imagenet/256/VIRTUAL_imagenet256_labeled.npz)？
Are the corresponding FID results reported in the paper calculated using adm_in256_stats.npz?
<img width="501" alt="截屏2024-08-02 14 40 49" src="https://github.com/user-attachments/assets/442baf03-901c-44cc-a6ce-44aeba70fb7e">

**Comment by Robootx** [comment]
same issue

**Comment by LTH14** [comment]
Thank you for your interest! Please note that the 10,000 images in the standard reference data of ADM's evaluation suite are NOT used for their FID evaluation. Instead, they use precomputed "mu" and "sigma" (also stored in the standard reference data) as the reference statistics of real images to compute the FID. Therefore, we have stored only the "mu" and "sigma" from the standard reference data into "adm_in256_stats.npz" and saved it directly in the github repo, so you don't need to download the large file of standard reference data as required by ADM. Our evaluation will generate the exact same result as the ADM evaluation suite. Please let me know if you have any other questions.

**Comment by MArSha1147** [comment]
Got it! Thank you for your detailed and clear answer!

**Comment by MArSha1147** [comment]
I hope this message finds you well.
I have a question regarding the reproduction of the rFID result in Table 2 of your paper. When I try to replicate your result for VAE kl-16 using the provided kl16.ckpt, I obtain an rFID value of 2.31, which is notably higher than the reported value of 1.43. I believe that the processes, including data transformation, latent sampling and the encoding/decoding procedures, are consistent with your code. Could you please confirm if there are any additional steps in the evaluation process that I may have overlooked? Alternatively, could it be possible that the kl16.ckpt released is not the optimal rFID version?
Thank you for your assistance! I look forward to your reply.

**Comment by LTH14** [comment]
I just double-checked the rFID by slightly modifying this codebase and the results are FID: 1.2150, Inception Score: 302.9750. Are you using the training or validation set to evaluate rFID? We used 50000 randomly sampled images from the ImageNet training set, using augmentation from DiT and ADM, and evaluated the rFID w.r.t. the statistics from the ADM suite.

**Comment by MArSha1147** [comment]
Thank you for your prompt assistance! I tested the kl16.ckpt using the reference data from ADM, just the same as adm_in256_stats.npz you provided. This could be the reason for the discrepancy.

**Comment by LTH14** [comment]
We also tested it w.r.t. adm_in256_stats.npz. Do you mean you use the images in the reference data from ADM to test rFID?

**Comment by MArSha1147** [comment]
Actually, the precomputed "mu" and "sigma" values from the [standard reference data](https://openaipublic.blob.core.windows.net/diffusion/jul-2021/ref_batches/imagenet/256/VIRTUAL_imagenet256_labeled.npz) were used to test rFID, and these values match those in your adm_in256_stats.npz file. I also tested it using the adm_in256_stats.npz and obtained the same rFID value of 2.31.

**Comment by LTH14** [comment]
Do you normalize the image data to [-1, 1]?

**Comment by MArSha1147** [comment]
Yes, I did it. Maybe there's still something going unnoticed here. Thx a lot for your kindness.

**Comment by Robot-Eyes** [comment]
For the provided kl16 VAE, I also got higher FID results with the ADM script and VIRTUAL_imagenet256_labeled.npz.
Inception Score: 213.57455444335938
FID: 2.3790881110223836
sFID: 3.90784467324022

The reconstruction code:

`
samples = samples.to(device, non_blocking=True)

    with torch.no_grad():
        posterior = vae.encode(samples)
        z = posterior.sample()
        recon_samples = vae.decode(z)
        recon_samples.clamp_(min=-1, max=1).sub_(-1).div_(2)
        recon_samples_np = recon_samples.mul(255).add_(0.5).clamp_(0, 255).permute(0, 2, 3, 1).to("cpu", torch.uint8).numpy()
`

**Comment by LTH14** [comment]
@Robootx we randomly sampled 50000 images from the ImageNet training set to compute the rFID

**Comment by Robot-Eyes** [comment]
Thanks for the info! I currently use
`transform_val = transforms.Compose([
    transforms.Lambda(lambda pil_image: center_crop_arr(pil_image, 256)),
    transforms.ToTensor(),
    transforms.Normalize([0.5, 0.5, 0.5], [0.5, 0.5, 0.5])
])`
When you said "the augmentation following DiT and ADM", did you mean
`transform_val = transforms.Compose([
    transforms.Resize((256, 256)),
    transforms.ToTensor(),
    transforms.Normalize([0.5, 0.5, 0.5], [0.5, 0.5, 0.5])
])`

**Comment by LTH14** [comment]
No, you shoud use the current one
```
transform = transforms.Compose([
    transforms.Lambda(lambda pil_image: center_crop_arr(pil_image, 256)),
    transforms.ToTensor(),
    transforms.Normalize([0.5, 0.5, 0.5], [0.5, 0.5, 0.5])
])
```

Do you randomly sample 50000 images from the training set to evaluate the rFID?

**Comment by Robot-Eyes** [comment]
The previous rFID=2.37 is from the validation set. I will try the training set next. Thanks for the info!

**Comment by LTH14** [comment]
rFID=2.37 from validation set is correct.

**Comment by Robot-Eyes** [comment]
> rFID=2.37 from validation set is correct -- this is because the FID is computed w.r.t. training statistics.

I see, that makes sense. Thanks for the prompt replies!

**Comment by Robot-Eyes** [comment]
> @Robootx we randomly sampled 50000 images from the ImageNet training set to compute the rFID, and used the augmentation following DiT and ADM. Could you please check whether you follow this recipe?

With this recipe (50k random training images regardless of class), I successfully reproduced the rFID similar to your Table 2:
Inception Score: 291.9043884277344
FID: 1.3728421406581788
sFID: 3.7944584578782496
